woocommerce/woocommerce-ios
# Product List: sort products based on name / date, ascending / descending

Fixes #2036 

- [x] ⚠️ update PR base to `develop` after https://github.com/woocommerce/woocommerce-ios/pull/2038 is merged ⚠️ 

## Changes

- Networking:
  - Added `date` case to the `ProductsRemote.OrderKey` enum to allow sorting by date
- Yosemite:
  - Created `ProductsSortOrder` enum for all different ways to sort products
  - For `ResultsController<StorageProduct>`, added a convenient initializer with a `ProductsSortOrder` param and a update function with a `ProductsSortOrder` param + unit tests
  - Added `ProductsSortOrder` to `ProductAction`'s `synchronizeProducts` and updated `ProductStore` to pass the corresponding `orderby` and `order` params to the API request + unit tests
- App:
  - Added a `UIAlertController` extension helper to present an action sheet with different sort options
    - Note for @Garance91540 and reviewer(s): there is a ✔️ on the current sort option in the design, but I couldn't find an official way to achieve this without modifying the text so leaving it out for this PR (I saw a [somewhat hacky way of showing the checkmark](https://stackoverflow.com/questions/40647039/how-to-add-uiactionsheet-button-check-mark) but felt like it could break with a future UIKit update)
  - In `ProductsViewController`, added a `ProductsSortOrder` field (default to ascending name) that can be updated from the action sheet above. Whenever the sort order changes, resync the first page of data (since different sort options could have different data in the first page, where the sort CTA is located)

## Testing

- Go to the Products tab
- Tap the "Sort by" CTA above the table view --> an action sheet should appear with 4 options
- Tap on "Title: A to Z" --> nothing should happen because this is the default
- Tap the "Sort by" CTA again --> an action sheet should appear with 4 options
- Tap on each of the sort option --> the table view should resync and display the first page of data based on the new sort option

## Example screenshots

light mode | dark mode
-- | --
![Simulator Screen Shot - iPhone 11 - 2020-03-26 at 15 46 19](https://user-images.githubusercontent.com/1945542/77622743-c9778280-6f79-11ea-8720-36b8d9699d30.png) | ![Simulator Screen Shot - iPhone 11 - 2020-03-26 at 15 52 15](https://user-images.githubusercontent.com/1945542/77622754-ce3c3680-6f79-11ea-851a-79e042594563.png)


Update release notes:
